From: Matthias Brugger <mbrugger@suse.com>
Subject: arm64: Disable kpti for non broadcast TLB HW

Patch-mainline: Never, this is suse internal to support older HW
References: bsc#1068032

Thunderx pass 1.1 HW has no broadcast TLB, which breaks kpti on this HW.
As the HW is not affected by any variant of Meltdown/Spectre we just disable ktpi early.

Signed-off-by: Matthias Brugger <mbrugger@suse.com>
---
 arch/arm64/kernel/cpufeature.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -691,6 +691,11 @@ static int __kpti_forced; /* 0: not forc
 static bool unmap_kernel_at_el0(const struct arm64_cpu_capabilities *entry,
 				int __unused)
 {
+        if (cpus_have_cap(ARM64_HAS_NO_BCAST_TLBI)) {
+                pr_info_once("HW does not has broadcast TLBI, disable KPTI\n");
+                return false;
+        }
+
 	/* Forced on command line? */
 	if (__kpti_forced) {
 		pr_info_once("kernel page table isolation forced %s by command line option\n",
